packages/reporters/hmr-server/src/HMRServer.js

```diff
@@ -0,0 +1,152 @@
+// @flow
+import type {ServerOptions, CacheEntry, Asset} from '@parcel/types';
+import type {PrintableError} from '@parcel/reporter-cli/src/prettyError';
+import type {Server, ServerError} from './types.js.flow';
+import http from 'http';
+import https from 'https';
+import WebSocket from 'ws';
+import generateCertificate from '@parcel/server-utils/src/generateCertificate';
+import getCertificate from '@parcel/server-utils/src/getCertificate';
+import logger from '@parcel/logger';
+import prettyError from '@parcel/reporter-cli/src/prettyError';
+
+type HMRMessage = {|
+  type: string,
+  error?: {
+    message: string,
+    stack?: string
+  },
+  assets?: Array<{
+    id: string,
+    type: string,
+    code: string,
+    deps: Object
+  }>
+|};
+
+export default class HMRServer {
+  server: Server;
+  wss: WebSocket.Server;
+  unresolvedError: HMRMessage | null = null;
+  changedAssets: Array<CacheEntry> = [];
+
+  async start(options: ServerOptions) {
+    await new Promise(async resolve => {
+      if (!options.https) {
+        this.server = http.createServer();
+      } else if (typeof options.https === 'boolean') {
```

Comment:
Nit: by the first condition not having been taken, this basically means `else if (options.https === true)`, which is a bit hard to read. This is totally personal preference, but what about:

```js
  if (options.https == null || options.https === false) {
    this.server = http.createServer();
  } else if (options.https === true) {
    this.server = https.createServer(generateCertificate(options));
  } else {
    // options.https is an object
    this.server = https.createServer(await getCertificate(options.https));
  }
```